lib/fd.js

```diff
@@ -0,0 +1,309 @@
+    this.creating = creating
+    this.appending = appending
+
+    this.position = null
+    this.blockPosition = stat ? stat.offset : null
+    this.blockOffset = 0
+
+    this._err = null
+    if (this.writable) {
+      if (this.appending) {
+        this._appendStream = this.drive.createReadStream(this.path)
+        this.position = this.stat.size
+      }
+    }
+
+    this._batcher = byteStream({ time: 100, limit: 4096 * 16 })
+
+    this._range = null
+  }
+
+  read (buffer, offset, len, pos, cb) {
+    if (!this.readable) return cb(new errors.BadFileDescriptor('File descriptor not open for reading.'))
+    if (this.position !== null && this.position === pos) this._read(buffer, offset, len, cb)
+    else this._seekAndRead(buffer, offset, len, pos, cb)
+  }
+
+  write (buffer, offset, len, pos, cb) {
+    if (!this.writable) return cb(new errors.BadFileDescriptor('File descriptor not open for writing.'))
+    if (!this.stat && !this.creating) {
+      return cb(new errors.BadFileDescriptor('File descriptor not open in create mode.'))
+    }
+    if (this.position !== null && pos !== this.position) {
+      return cb(new errors.BadFileDescriptor('Random-access writes are not currently supported.'))
+    }
+    if (this.appending && pos < this.stat.size) {
+      return cb(new errors.BadFileDescriptor('Position cannot be less than the file size when appending.'))
+    }
+    if (!this._writeStream && !this.appending && pos) {
+      return cb(new errors.BadFileDescriptor('Random-access writes are not currently supported.'))
+    }
+
+    const self = this
+    if (!this._writeStream) {
+      const writeStream = this.drive.createWriteStream(this.path)
+      this._writeStream = pumpify(this._batcher, through.obj((chunk, enc, cb) => {
+        cb(null, Buffer.concat(chunk)) 
+      }), writeStream)
+      this._writeStream.on('error', err => {
+        return cb(err)
+      })
+    }
+
+    // TODO: This is a temporary (bad) way of supporting appends.
+    if (this._appendStream) {
+      this.position = this.stat.size
+      // pump does not support the `end` option.
+      this._appendStream.pipe(this._writeStream, { end: false })
+
+      this._appendStream.on('error', err => this._writeStream.destroy(err))
+      this._writeStream.on('error', err => this._appendStream.destroy(err))
+
+      return this._appendStream.on('end', doWrite)
+    }
+
+    return doWrite()
+
+    function doWrite (err) {
+      self._appendStream = null
+      if (err) return cb(err)
+      if (self._err) return cb(self._err)
+      if (self._writeStream.destroyed) return cb(new errors.BadFileDescriptor('Write stream was destroyed.'))
+
+      const slice = buffer.slice(offset, len)
+      write(self._writeStream, slice, err => {
+        if (err) return cb(err)
+        self.position += slice.length
+        return cb(null, slice.length, buffer)
+      })
+    }
+
+    function write(stream, data, cb) {
+      if (!stream.write(data)) {
```

Comment:
to be super streamy you actually have to do stream.write(data) === false for this check